pt/endpoints/powerdata/resource.py

```diff
@@ -144,148 +148,157 @@ def data_table(self, limit, offset, time, field):
+        # Demand 7 days power usage
+        demand_list = db.session.query(
+            cast(Demand.updated_at + func.cast(concat(8, ' HOURS'), INTERVAL), DATE).label('date'),
+            (func.sum(Demand.grid) / 60).label('sum')
+        ).filter(
+            Demand.uuid.in_(
+                [data.uuid for data in PowerData.query.filter(
+                    PowerData.updated_at.between(start_time, end_time),
+                    PowerData.field == field,
+                    PowerData.data_type == 'Demand'
+                ).all()]
+            )
+        ).group_by('date').order_by('date').all()
+        # PV 7 days power usage
+        pv_list = db.session.query(
+            cast(PV.updated_at + func.cast(concat(8, ' HOURS'), INTERVAL), DATE).label('date'),
+            (func.sum(PV.pac) / 60).label('sum')
+        ).filter(
+            PV.uuid.in_(
+                [data.uuid for data in PowerData.query.filter(
+                    PowerData.updated_at.between(start_time, end_time),
+                    PowerData.field == field,
+                    PowerData.data_type == 'PV'
+                ).all()]
+            )
+        ).group_by('date').order_by('date').all()
+        # EV 7 days power usage
+        ev_list = db.session.query(
+            cast(EV.updated_at + func.cast(concat(8, ' HOURS'), INTERVAL), DATE).label('date'),
+            (func.sum(EV.power_display) / 60).label('sum')
+        ).filter(
+            EV.uuid.in_(
+                [data.uuid for data in PowerData.query.filter(
+                    PowerData.updated_at.between(start_time, end_time),
+                    PowerData.field == field,
+                    PowerData.data_type == 'EV'
+                ).all()]
+            )
+        ).group_by('date').order_by('date').all()
+        # ESS 7 days power usage
+        ess_list = db.session.query(
+            cast(ESS.updated_at + func.cast(concat(8, ' HOURS'), INTERVAL), DATE).label('date'),
+            (func.sum(ESS.power_display) / 60).label('sum')
+        ).filter(
+            ESS.uuid.in_(
+                [data.uuid for data in PowerData.query.filter(
+                    PowerData.updated_at.between(start_time, end_time),
+                    PowerData.field == field,
+                    PowerData.data_type == 'ESS'
+                ).all()]
+            )
+        ).group_by('date').order_by('date').all()
+        # WT 7 days power usage
+        wt_list = db.session.query(
+            cast(WT.updated_at + func.cast(concat(8, ' HOURS'), INTERVAL), DATE).label('date'),
+            (func.sum(WT.windgridpower) / 60).label('sum')
+        ).filter(
+            WT.uuid.in_(
+                [data.uuid for data in PowerData.query.filter(
+                    PowerData.updated_at.between(start_time, end_time),
+                    PowerData.field == field,
+                    PowerData.data_type == 'WT'
+                ).all()]
+            )
+        ).group_by('date').order_by('date').all()
```

Comment:
First, why not use last suggestion of the query?
Cost time of querying 5 types of data:
+ old way: 0:00:04.620344
+ new way: 0:00:00.876924
It's 2.7x more faster!
```
db.session.query(cast(PV.updated_at + func.cast(concat(8, ' HOURS'), INTERVAL), DATE).label('date'), func.sum(PV.pac).label('sum')).filter(PV.updated_at.between(start, end), PV.field == "SGESC_C_BEMS").group_by('date').order_by('date').all()
```
Filter by uuid is not needed, because the data_type of tables are inherits from `PowerData`, the query of field in `PowerData` can directly use by the five data_type table.

Second, the query is similar, only some variables change. You can try to simplify them.
For example, put variables into a dictionary, use for-loop to query.
eg.
```
powerdata_datatype = {
    "pv": {
        "model": PV,
        "field": PV.pac,
    },
    "demand": {
        "model": Demand,
        "field": Demand.grid,
    }
    ...
}
powerdata_sum = {}
for data_type in powerdata_datatype:
    powerdata_sum[data_type] = db.session.query(cast(powerdata_datatype[data_type]["model"].updated_at + func.cast(concat(8, ' HOURS'), INTERVAL), DATE).label('date'), func.sum(powerdata_datatype[data_type]["field"]).label('sum')).filter(PV.updated_at.between(start, end), powerdata_datatype[data_type]["model"].field == "SGESC_C_BEMS").group_by('date').order_by('date').all()
```
If you have time, can try SQL JOIN. If you are busy, can just put JOIN into issue.

Reply:
Noted, thanks for the implementation.
I did try to use the query you provided, but I did not catch up with the inheritance of PowerData Model and related DataType Model. In the last commit I had tried to reduce the time from 8 seconds to 4.5 seconds...

Anyways, thanks, the work of advance querying is beautiful, I'll improve it.

Reply:
I have put JOIN into issue, you can use the code above and it's OK.